### PR TITLE
Feature/cs 282 common go money

### PR DESCRIPTION
**Pull request by jcShopline**

### CODE REVIEW

File: money.go [hunk 1]
```diff
@@ -107,53 +166,78 @@ func (m *Money) Negative() *Money {
...
-	m.initMoney()
-	om.initMoney()
-
-	nm, err := m.money.Add(om.money)
-	if err != nil {
-		return nil, err
+func (m *Money) Add(oms ...*Money) (*Money, error) {
+	result := m
+	for _, om := range oms {
+		result.initMoney()
```
Comment (andy-shopline) [inline money.go]: result being inited many times inside the loop

File: money.go [hunk 2]
```diff
@@ -107,53 +166,78 @@ func (m *Money) Negative() *Money {
...
+		om.initMoney()
+		nm, err := result.money.Add(om.money)
+		if err != nil {
+			return nil, err
+		}
+		result = &Money{
+			Cents:          nm.Amount(),
+			Dollars:        nm.AsMajorUnits(),
+			CurrencyIso:    m.CurrencyIso,
+			CurrencySymbol: m.getRoundingModeAmongMoneys(oms),
```
Comment (andy-shopline) [inline money.go]: why currency symbol becomes rounding mode...

File: money.go [hunk 3]
```diff
@@ -107,53 +166,78 @@ func (m *Money) Negative() *Money {
...
+		if err != nil {
+			return nil, err
+		}
+
+		result = &Money{
+			Cents:          nm.Amount(),
+			Dollars:        nm.AsMajorUnits(),
+			CurrencyIso:    m.CurrencyIso,
+			CurrencySymbol: m.CurrencySymbol,
+			roundingMode:   m.getRoundingModeAmongMoneys(oms),
```
Comment (andy-shopline) [inline money.go]: should we assign the rounding mode after the loop?

File: money.go [hunk 4]
```diff
@@ -4,6 +4,18 @@ import (
 	"math"
 
 	gomoney "github.com/Rhymond/go-money"
+	"github.com/samber/lo"
+)
+
+func init() {
+	// Need to change Currency TWD Fraction from 2 to 0 in /Rhymond/go-money
+	gomoney.AddCurrency("TWD", "NT$", "$1", ".", ",", 0)
```
Comment (Seitk) [inline money.go]: move to `currency.go` file

Comment (Seitk) [reply money.go]: also add other currencies

File: money.go [hunk 5]
```diff
@@ -4,6 +4,18 @@ import (
...
+}
+
+const (
+	RoundUp     = "ROUND_UP"
+	RoundDown   = "ROUND_DOWN"
+	BankerRound = "ROUND_HALF_EVEN"
```
Comment (Seitk) [inline money.go]: better to keep the constant name align, do `RoundHalfEven` or `RoundBankers`

File: money.go [hunk 6]
```diff
@@ -13,26 +25,72 @@ type Money struct {
 	Label          string  `json:"label" bson:"label"`
 	Dollars        float64 `json:"dollars" bson:"dollars"`
 
-	money *gomoney.Money
+	roundingMode string
+	money        *gomoney.Money
 }
 
-func New(cents int64, isoCode string) *Money {
+func New(cents int64, isoCode string, roundingMode string) *Money {
```
Comment (Seitk) [inline money.go]: Use option pattern to shadow optional argument

```
type MoneyOption func(*Money)

func WithRoundingMode(mode string) MoneyOption {
	return func(m *Money) {
		m.roundingMode = mode
	}
}

func New(cents int64, isoCode string, options ...MoneyOption) *Money {
  ..
  money := &Money{
   ...
  }
  ...
  for _, option := range options {
    option(money)
  }
}
```

File: money.go [hunk 7]
```diff
@@ -13,26 +25,72 @@ type Money struct {
...
+}
+
 func (m *Money) initMoney() {
 	if m.money == nil {
 		m.money = gomoney.New(m.Cents, m.CurrencyIso)
 	}
 }
 
+// Getting the roound mode among all Money. If om is not exist, will return the first non-nil round mode. Otherwise will return ""
+func (m *Money) getRoundingModeAmongMoneys(ma []*Money) string {
```
Comment (Seitk) [inline money.go]: maybe rename to `alignRoundingMode` and re-assign `m.roundingMode` within this function

File: money.go (hunk 6 above)
Comment (Seitk) [reply money.go]: btw can add `RoundBankers` as default in `New` function

File: money.go [hunk 8]
```diff
@@ -13,26 +25,72 @@ type Money struct {
...
+	})
+	if isFound {
+		return fm.roundingMode
+	}
+	return ""
+
+}
+
+// Rounded function. Default mode will be Banker rounding mode
+func (m *Money) roundingByModeSet(value float64) float64 {
```
Comment (Seitk) [inline money.go]: rename to `RoundByMode`, I think is fine to public this function if we need to manually reuse the logic outside

File: money_test.go [hunk 9]
```diff
@@ -1,41 +1,935 @@
...
+		{
+			cents:    100,
+			currency: "HKD",
+			expected: "$1.00",
+		},
+		{
+			cents:    100,
+			currency: "USD",
+			expected: "$1.00",
+		},
```
Comment (Seitk) [inline money_test.go:255]: add all the currencies we supported

File: money.go [hunk 10]
```diff
@@ -107,53 +166,80 @@ func (m *Money) Negative() *Money {
...
 
 // Add returns new Money struct with value representing sum of Self and Other Money.
-func (m *Money) Add(om *Money) (*Money, error) {
-	m.initMoney()
-	om.initMoney()
-
-	nm, err := m.money.Add(om.money)
-	if err != nil {
-		return nil, err
+func (m *Money) Add(oms ...*Money) (*Money, error) {
```
Comment (andy-shopline) [inline money.go:201]: can refactor add like this. just reuse the embedded money object to do add operation and wrap it in the return
```go
func (m *Money) Add(oms ...*Money) (*Money, error) {
	m.initMoney()
	innerMoney := m.money
	var err error
	for _, om := range oms {
		om.initMoney()
		innerMoney, err = innerMoney.Add(om.money)
		if err != nil {
			return nil, err
		}
	}
	return &Money{
		Cents:          innerMoney.Amount(),
		Dollars:        innerMoney.AsMajorUnits(),
		CurrencyIso:    m.CurrencyIso,
		CurrencySymbol: m.CurrencySymbol,
		roundingMode:   m.getRoundingModeAmongMoneys(oms),
	}, nil
}
```

File: money.go [hunk 11]
```diff
@@ -13,26 +25,72 @@ type Money struct {
...
 		m.money = gomoney.New(m.Cents, m.CurrencyIso)
 	}
 }
 
+// Getting the roound mode among all Money. If om is not exist, will return the first non-nil round mode. Otherwise will return ""
+func (m *Money) getRoundingModeAmongMoneys(ma []*Money) string {
+	if m.roundingMode != "" {
+		return m.roundingMode
+	}
+	fm, _, isFound := lo.FindIndexOf(ma, func(money *Money) bool {
```
Comment (andy-shopline) [inline money.go]: you can use lo.Find if you don't want the index

File: money_test.go [hunk 12]
```diff
@@ -1,41 +1,935 @@
...
+	for _, item := range testTable {
+		m2 := New(int64(item.cents), item.currency, RoundDown)
+		nm, err := m1.Equals(m2)
+		assert.NoError(t, err)
+		assert.Equal(t, item.expected, nm)
+
+	}
+}
+
+func TestEqual_DifferentCurrenties(t *testing.T) {
```
Comment (andy-shopline) [inline money_test.go]: currencies

**Review by Seitk — APPROVED** [review]

**Review by Seitk — APPROVED** [review]

**Review by andy-shopline — APPROVED** [review]

**Pull request merged**